source/magpie/dataframe.d

```diff
@@ -93,16 +149,13 @@ public:
                 {
                     size_t tmp = 0;
                     if(indx.row.codes[i].length == 0)
-                        tmp = indx.row.index[i][0 .. top - indx.column.index.length - extra].map!(e => e.length).reduce!max;
+                        tmp = gapCalc(indx.row.index[i][0 .. top - indx.column.index.length - extra]);
                     else if(indx.row.index[i].length == 0)
-                        tmp = indx.row.codes[i][0 .. top - indx.column.index.length - extra].map!(e => to!string(e).length).reduce!max;
+                        tmp = gapCalc(indx.row.codes[i][0 .. top - indx.column.index.length - extra]);
                     else
                         tmp = indx.row.codes[i][0 .. top - indx.column.index.length - extra].map!(e => indx.row.index[i][e].length).reduce!max;
 
-                    if(tmp > thisGap)
-                    {
-                        thisGap = tmp;
-                    }
+                    thisGap = max(thisGap, tmp);
                 }
```

Comment:
You have this same block of code repeated three times with different different indices, you should make that a (nested) function.